### ISSUE TITLE
Support arm architecture on macOS

### ISSUE THREAD

**Issue by tisoft** [LWJGLX/lwjgl3-awt#37]
The native library needs to be compiled with arm support.

Depends on https://github.com/LWJGL/lwjgl3/issues/601

**Comment by tisoft** [comment]
LWJGL switches from dyncall to libffi, which should allow us to get rid of the native library altogether.

**Comment by SWinxy** [comment]
@tisoft I messed around with Libffi, and we would just need the function pointer to the MetalKit methods. I'm not sure how this would be done. It's probably something simple though.

**Comment by brentco** [comment]
Any progress on this? I was trying to rebuild the library with the provided command but it's missing JavaVM. Apparently there's some weird stuff going on with that since a certain xcode version (don't ask me...). I had to remove the "JavaVM/" prefixes in every `#include` and then I was able to compile with the following command: 
```bash
gcc -I/Library/Java/JavaVirtualMachines/temurin-17.jdk/Contents/Home/include -I/Library/Java/JavaVirtualMachines/temurin-17.jdk/Contents/Home/include/darwin -dynamiclib lwjgl3awt/*.m -o native/macosx/liblwjgl3awt.dylib -framework CoreFoundation -framework AppKit -framework MetalKit -framework Metal
``` 
Of course the jdk location is specific to my installation, so replace with your own.

I don't even know if that will work, we'll see :)

**Comment by brentco** [comment]
> @tisoft I messed around with Libffi, and we would just need the function pointer to the MetalKit methods. I'm not sure how this would be done. It's probably something simple though.

I briefly checked this tutorial about libffi https://eli.thegreenplace.net/2013/03/04/flexible-runtime-interface-to-shared-libraries-with-libffi and I noticed they locate the function pointer by name with `dlsym(libHandle, functionName)`. Is this useful perhaps?

**Comment by SWinxy** [comment]
I asked @Spasi on the discord on what to do. Using `MacOSXLibrary.getWithIdentifier("com.apple.MetalKit")` gets a valid library, from which you can call `getAddressPtr()` (which seems equivalent to `dlsym`). Spasi also told me that the things used in the native code are just normal C functions. But calling `getAddressPtr` with a method name returns `NULL`. I still am not sure how to do this.

**Comment by brentco** [comment]
> 

Briefly looking into the source code I found this: https://github.com/LWJGL/lwjgl3/blob/e4a6cc863f469ea8acfe3c2158f2c77d0c0aa95d/modules/lwjgl/core/src/main/java/org/lwjgl/system/FunctionProvider.java#L16 which is an interface that MacOSXLibrary implements. If it's a function you are looking for, then I believe `getFunctionAddress(CharSequence functionName)` might be more suitable?

**Comment by SWinxy** [comment]
`MacOSXLibrary.getFunctionAddress(CharSequence functionName)` is what I was thinking of. Too lazy to have looked it up.

**Comment by Spasi** [comment]
@SWinxy This works for me:

```
SharedLibrary lib = MacOSXLibrary.create("/System/Library/Frameworks/Metal.framework");
System.out.println(lib.getFunctionAddress("MTLCreateSystemDefaultDevice"));
```

However, Metal is mostly a Swift/ObjC API and `dlsym` won't help. You'll need the Objective-C Runtime bindings to use Metal.

**Comment by skalarproduktraum** [comment]
Hi everyone! I got the library to compile on ARM, and also a (somewhat eviscerated) version of the TriangleDemo (such that it only clears the screen to cornflower blue) to work on my M1 MBP:

<img width="790" alt="Screenshot 2022-04-12 at 20 02 57" src="https://user-images.githubusercontent.com/586495/163025708-f580085d-83be-4e84-b0d4-9a4abf77569a.png">

I wonder if this is useful to anyone? Maybe @tisoft or @httpdigest can comment? The native library is at the moment linked to a Temurin 17 install, I guess it'd also work with 11. I have also played around with the ObjC wrapper, but I got stuck at the CGMakeRect call, which is apparently inlined, so it doesn't have an associated symbol. Let me know if there's anything I can do or test, having this fully working on M1 would be great for several of my projects 👍

**Comment by skalarproduktraum** [comment]
In the meantime, I've also managed to create a universal binary for AMD64 and ARM64:

```
➜  lwjgl3-awt git:(main) ✗ gcc -dynamiclib lwjgl3awt/*.m -o native/macosx/liblwjgl3awt-arm64.dylib -framework CoreFoundation -framework AppKit -framework MetalKit -framework Metal -I~/Library/Java/JavaVirtualMachines/temurin-11.0.14.1/Contents/Home/include/darwin/ -I~/Library/Java/JavaVirtualMachines/temurin-11.0.14.1/Contents/Home/include/

[deprecation warnings omitted]
➜  lwjgl3-awt git:(main) ✗ gcc -arch x86_64 -dynamiclib lwjgl3awt/*.m -o native/macosx/liblwjgl3awt-x86_64.dylib -framework CoreFoundation -framework AppKit -framework MetalKit -framework Metal -I~/Library/Java/JavaVirtualMachines/temurin-11.0.14.1/Contents/Home/include/darwin/ -I ~/Library/Java/JavaVirtualMachines/temurin-11.0.14.1/Contents/Home/include/

[deprecation warnings omitted]
➜  lwjgl3-awt git:(main) ✗ file native/macosx/*
native/macosx/liblwjgl3awt-arm64.dylib:  Mach-O 64-bit dynamically linked shared library arm64
native/macosx/liblwjgl3awt-x86_64.dylib: Mach-O 64-bit dynamically linked shared library x86_64
native/macosx/liblwjgl3awt.dylib:        Mach-O 64-bit dynamically linked shared library arm64
➜  lwjgl3-awt git:(main) ✗ lipo -create native/macosx/liblwjgl3awt-arm64.dylib native/macosx/liblwjgl3awt-x86_64.dylib -
output native/macosx/liblwjgl3awt-universal.dylib
➜  lwjgl3-awt git:(main) ✗ file native/macosx/liblwjgl3awt-universal.dylib
native/macosx/liblwjgl3awt-universal.dylib: Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit dynamically linked shared library x86_64Mach-O 64-bit dynamically linked shared library x86_64] [arm64:Mach-O 64-bit dynamically linked shared library arm64Mach-O 64-bit dynamically linked shared library arm64]
native/macosx/liblwjgl3awt-universal.dylib (for architecture x86_64):	Mach-O 64-bit dynamically linked shared library x86_64
native/macosx/liblwjgl3awt-universal.dylib (for architecture arm64):	Mach-O 64-bit dynamically linked shared library arm64
```

This and the previous attempt were actually also linked against Temurin 11, not 17, as stated above.

This looks like it works, haven't been able to test this on an AMD64 Mac, though.

**Comment by Hc747** [comment]
@skalarproduktraum is there any way we could get the apple silicon binaries for Java 8?

**Comment by skalarproduktraum** [comment]
Hey @Hc747! Have a look at #57, this is a new version of my PR that actually does not require binaries anymore 👍 Let me know if you need any help trying this out!

**Comment by SWinxy** [comment]
Solved in #57.
